src/v/test_utils/archival.h

```diff
@@ -98,3 +100,60 @@ inline void populate_manifest(
              spec.end_offset - spec.end_kafka_offset)});
     }
 }
+
+inline archival::upload_candidate_with_locks
+require_upload_candidate(archival::candidate_creation_result&& r) {
+    ss::visit(
+      r,
+      [](std::monostate) {
+          BOOST_FAIL("unexpected default candidate creation result");
+      },
+      [](const archival::candidate_creation_error& err) {
+          BOOST_FAIL(fmt::format("unexpected creation error: {}", err));
+      },
+      [](const archival::skip_offset_range& r) {
+          BOOST_FAIL(fmt::format("unexpected skip offset range: {}", r.reason));
+      },
+      [](const archival::upload_candidate_with_locks&) {});
+    return std::move(std::get<archival::upload_candidate_with_locks>(r));
+}
+
+inline void require_candidate_creation_error(
+  archival::candidate_creation_result&& r,
+  archival::candidate_creation_error expected) {
+    ss::visit(
+      r,
+      [](std::monostate) {
```

Comment:
i see. i wonder if this is better / worse than `std::optional<std::variant<...>>`

Reply:
IMO, monostate is better because the caller is forced to handle its presence when visiting the states of the variant, the compiler will warn if the visit is non-exhaustive, and monostate indicates that some code path was taken which is unexpected. 

With an optional at the call site (IE the caller using `optional<res>` initialized to nullopt to store the result) the caller is expected to use this extra step and then check it, and all future call sites also have to remember to do this.

If we return `optional<res>` directly from the segment collector/archiver policy methods then those methods look a little difficult to understand as nothing in the methods will ever return nullopt but the return type is now optional. 

With monostate the call site is a little more verbose but the type contains enough details for the caller to be able to indicate a problematic state.

But if there are arguments against `monostate` which can cause issues please let me know. 

If there is a way in C++ to treat conditional branches as expressions it would be very useful here.

Reply:
Agree with @abhijat on this one. The only best next thing would be to remove default constructors from the members of the variant. Something we should consider doing but it would have been too much probably in this PR.

Reply:
Cool, thanks for the explanation. I agree now too!